packages/react-components/react-dialog/src/components/Dialog/useDialog.ts

```diff
@@ -9,20 +15,177 @@ import type { DialogProps, DialogState } from './Dialog.types';
  * before being passed to renderDialog_unstable.
  *
  * @param props - props from this instance of Dialog
- * @param ref - reference to root HTMLElement of Dialog
  */
-export const useDialog_unstable = (props: DialogProps, ref: React.Ref<HTMLElement>): DialogState => {
+export const useDialog_unstable = (props: DialogProps): DialogState => {
+  const { children, overlay, modalType = 'modal', onOpenChange } = props;
+
+  const [trigger, content] = childrenToTriggerAndContent(children);
+
+  const [open, setOpen] = useControllableState({
+    state: props.open,
+    defaultState: props.defaultOpen,
+    initialState: false,
+  });
+
+  const overlayShorthand = resolveShorthand(overlay, {
+    required: modalType !== 'non-modal',
+    defaultProps: {
+      'aria-hidden': 'true',
+    },
+  });
+
+  const requestOpenChange = useEventCallback((data: DialogRequestOpenChangeData) => {
+    const getNextOpen = normalizeSetStateAction(data.open);
+    const isDefaultPrevented = normalizeDefaultPrevented(data.event);
+
+    if (onOpenChange) {
+      onOpenChange(data.event, {
```

Comment:
should `onOpenChange` also be called if the user prevents  default ?

Reply:
How will the user prevent default before `props.onOpenChange` gets called?

Reply:
I would never have thought `onOpenChange` would be the way to prevent default. Since in that case the user could just as easily control the component and get it to not open ?

I thought this would handle prevent default by explicit handlers. Although now when I think of this it seems quite complex and I can't see a clear scenario where prevent default should be handled like this. Can you provide an example ?

Reply:
On explicit handler, you could simply not call `onOpenChange` instead of preventing default.

The way I see preventing default is a signature between the user and the library, where you state a behavior and you let the user manage this mechanism to prevent this behavior. One case that comes to mind is preventing the dialog from closing on clicking on determined part of the overlay, let's say a logo or something for example.